HTTP server do SSL handshake in client fiber

waj opened the pull request:
This fixes #8108 by forcing the call to `SSL_accept` within the fiber that handles the request.

To implement this fix, I added the `OpenSSL::SSL::Server#start_immediately` property just like Ruby does: https://ruby-doc.org/stdlib-2.5.1/libdoc/openssl/rdoc/OpenSSL/SSL/SSLServer.html

straight-shoota commented on src/openssl/ssl/server.cr:
```diff
@@ -60,18 +63,24 @@ class OpenSSL::SSL::Server
     end
   end
 
+  private def new_server(io)
```
I'd rather call this `new_server_socket` or `new_ssl_socket`. It doesn't create a server (`self` is already the server).

waj replied:
Done

Sija commented on src/openssl/ssl/socket.cr:
```diff
@@ -53,10 +53,13 @@ abstract class OpenSSL::SSL::Socket < IO
         ret = LibSSL.ssl_accept(@ssl)
         unless ret == 1
-          io.close if sync_close
+          @bio.io.close if @sync_close
```
```suggestion
          @bio.io.close if sync_close?
```

Sija replied:
```suggestion
          @bio.io.close if sync_close?
```

Sija commented on src/openssl/ssl/server.cr:
```diff
@@ -35,6 +35,9 @@ class OpenSSL::SSL::Server
   getter? closed : Bool = false
 
+  # When `true` a call to `#accept` will also initiate the SSL handshake.
+  property start_immediately : Bool = true
```
```suggestion
  property? start_immediately : Bool = true
```

waj replied:
I don't think every boolean property must have a question mark. In particular I like to use it when is a read only property or with methods. Also, I copied this name from Ruby's OpenSSL and I prefer to keep the same name.

Sija commented on src/openssl/ssl/server.cr:
```diff
@@ -60,18 +63,24 @@ class OpenSSL::SSL::Server
 
+  private def new_server(io)
+    server = OpenSSL::SSL::Socket::Server.new(io, @context, sync_close: @sync_close)
+    server.accept if @start_immediately
```
```suggestion
    server.accept if start_immediately?
```

asterite replied:
This is not needed. A call is a bit more expensive for the compiler in terms of compiling.

asterite replied:
Put another way: both do the same thing and we don't prefer one over the other.

Sija replied:
I'd argue that in the vein of OOP we should prefer using getters over ivars (ability to overload 'em by subclasses)

asterite replied:
I then call YAGNI

waj commented:
Just added a spec and fixed the segfaults due to a double SSL_free call. Because now the `accept` is not called within the constructor, the finalizer will be executed correctly.

RX14 commented on src/openssl/ssl/socket.cr:
```diff
@@ -53,15 +53,13 @@ abstract class OpenSSL::SSL::Socket < IO
-        raise ex
+    end
+
+    def accept
```
Isn't this a breaking change?

waj replied:
That's true. I just refactored this fix so it doesn't break compatibility. Thanks!

waj commented:
@RX14 I agree with you, like a do with @straight-shoota. I just did the smallest change to fix this bug with the current design, but I'm also with you in that these wrappers around OpenSSL api is not ideal.

636f7374 commented:
Great!
